File: benchmarks/cross-lang/async-echo/go/bench.go
// Phase 31: async-echo (Go)
// Contract: fanout_fanin_real_concurrency.v2.

package main

import (
	"fmt"
	"os"
	"sync"
)

func main() {
	const iters = 10000
	total := 0
	for i := 0; i < iters; i++ {
		var wg sync.WaitGroup
		var local int64
		var mu sync.Mutex
		for k := 1; k <= 10; k++ {
			wg.Add(1)
			go func(v int) {
				defer wg.Done()
				mu.Lock()
				local += int64(v)
				mu.Unlock()
			}(k)
		}
		wg.Wait()
		total += int(local)
	}
	if total != 550000 {
		fmt.Fprintf(os.Stderr, "unexpected: %d\n", total)
		os.Exit(1)
	}
}

File: benchmarks/cross-lang/async-echo/rust/bench.rs
// Phase 31: async-echo (Rust)

use std::sync::{Arc, Mutex};
use std::thread;

fn main() {
    let iters: usize = 10000;
    let mut total: i64 = 0;
    for _ in 0..iters {
        let local = Arc::new(Mutex::new(0_i64));
        let mut handles = Vec::with_capacity(10);
        for k in 1..=10 {
            let local = Arc::clone(&local);
            handles.push(thread::spawn(move || {
                let mut g = local.lock().unwrap();
                *g += k as i64;
            }));
        }
        for h in handles {
            h.join().unwrap();
        }
        let g = local.lock().unwrap();
        total += *g;
    }
    if total != 550_000 {
        eprintln!("unexpected: {}", total);
        std::process::exit(1);
    }
}
